### PR TITLE
SoundWire: fix port_ready[] allocation and remove it when unnecessary

### PR DESCRIPTION
**Pull request by plbossart**
The use of port_ready[] is broken for ports that are not contiguous, and none of the existing codec devices use this functionality anyways. Let's fix the core and remove unnecessary allocations in codec drivers.

No obvious regressions seen on CML-U w/ SoundWire, but additional checks might not hurt.

@shumingfan @ryans-lee @oder-chiou @jack-cy-yu Can you do me a favor and double-check that the use of the simplified channel prepare state machine is indeed a feature of your hardware, not a copy/paste error or omission. If we do need to enable the more complicated state machine then the last patch needs to be redone. Thanks!

### CODE REVIEW

**Comment by bardliao** [comment]
@plbossart Now slave->port_ready is only allocated and initialized at sdw_slave_read_prop(). But I don't see sdw_slave_read_prop() is called anywhere. Will it cause problem? Can we move the allocation and initialization to sdw_slave_add() since it is a fixed array now?

**Comment by RanderWang** [comment]
> @plbossart Now slave->port_ready is only allocated and initialized at sdw_slave_read_prop(). But I don't see sdw_slave_read_prop() is called anywhere. Will it cause problem? Can we move the allocation and initialization to sdw_slave_add() since it is a fixed array now?

I also found this. I added log in sdw driver but found port_ready was not checked by complete(&slave->port_ready[port]) on TGL RVP for no port irq was triggered.

**Comment by plbossart** [comment]
> > @plbossart Now slave->port_ready is only allocated and initialized at sdw_slave_read_prop(). But I don't see sdw_slave_read_prop() is called anywhere. Will it cause problem? Can we move the allocation and initialization to sdw_slave_add() since it is a fixed array now?
> 
> I also found this. I added log in sdw driver but found port_ready was not checked by complete(&slave->port_ready[port]) on TGL RVP for no port irq was triggered.

Sorry I don't get your point @RanderWang. If the codecs uses the simplified channel prepare state machine, there will not be any port_ready interrupts.

**Comment by plbossart** [comment]
> @plbossart Now slave->port_ready is only allocated and initialized at sdw_slave_read_prop(). But I don't see sdw_slave_read_prop() is called anywhere. Will it cause problem? Can we move the allocation and initialization to sdw_slave_add() since it is a fixed array now?

we could indeed move to a common place, good suggestion, so that codec drivers don't need to worry about this.

**Comment by RanderWang** [comment]
> > > @plbossart Now slave->port_ready is only allocated and initialized at sdw_slave_read_prop(). But I don't see sdw_slave_read_prop() is called anywhere. Will it cause problem? Can we move the allocation and initialization to sdw_slave_add() since it is a fixed array now?
> > 
> > 
> > I also found this. I added log in sdw driver but found port_ready was not checked by complete(&slave->port_ready[port]) on TGL RVP for no port irq was triggered.
> 
> Sorry I don't get your point @RanderWang. If the codecs uses the simplified channel prepare state machine, there will not be any port_ready interrupts.

I meant not error in original PR for the port_ready was not checked by slave, or it would lead to kernel panic.
That was why Bard said it would cause problem.

File: drivers/soundwire/mipi_disco.c [hunk 1]
```diff
@@ -289,7 +289,7 @@ int sdw_slave_read_prop(struct sdw_slave *slave)
 	struct sdw_slave_prop *prop = &slave->prop;
 	struct device *dev = &slave->dev;
 	struct fwnode_handle *port;
-	int num_of_ports, nval, i, dp0 = 0;
+	int nval, i, dp0 = 0;
```
Comment (RanderWang) [inline drivers/soundwire/mipi_disco.c]: build warning
````
drivers/soundwire/mipi_disco.c: In function ‘sdw_slave_read_prop’:
drivers/soundwire/mipi_disco.c:292:12: warning: unused variable ‘i’ [-Wunused-variable]
  int nval, i, dp0 = 0;
````

**Comment by RanderWang** [comment]
Tested on TGL RVP. no regression issue was found.

File: drivers/soundwire/slave.c [hunk 2]
```diff
@@ -25,11 +26,18 @@ static int sdw_slave_add(struct sdw_bus *bus,
 {
 	struct sdw_slave *slave;
 	int ret;
+	int i;
 
 	slave = kzalloc(sizeof(*slave), GFP_KERNEL);
 	if (!slave)
 		return -ENOMEM;
 
+	slave->port_ready = kcalloc(SDW_MAX_PORTS, sizeof(*slave->port_ready), GFP_KERNEL);
```
Comment (lyakh) [inline drivers/soundwire/slave.c]: then we could go a step further and embed the array into `struct sdw_slave`?

File: sound/soc/codecs/max98373-sdw.c [hunk 3]
```diff
@@ -333,17 +331,6 @@ static int max98373_read_prop(struct sdw_slave *slave)
 		i++;
 	}
 
-	/* Allocate port_ready based on num_of_ports */
-	slave->port_ready = devm_kcalloc(&slave->dev, num_of_ports,
-					 sizeof(*slave->port_ready),
-					 GFP_KERNEL);
```
Comment (lyakh) [inline sound/soc/codecs/max98373-sdw.c:342]: am I confused or does this mean, that between the previous and this patch `slave->port_ready` was assigned by two allocations? Maybe we can avoid this somehow... E.g. if we embed the array and are forced to merge these two patches?

File: drivers/soundwire/slave.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/slave.c]: yes it's true. Will make that change, thanks for the suggestion.

File: sound/soc/codecs/max98373-sdw.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/codecs/max98373-sdw.c:342]: the allocation was initially done either in the mipi_disco stuff, or in the codec driver if it didn't use the mipi_disco stuff.
The point now is that we remove both alternatives and do a single allocation in a single place

File: drivers/soundwire/slave.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/slave.c]: done

File: sound/soc/codecs/max98373-sdw.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/codecs/max98373-sdw.c:342]: fixed now

**Comment by plbossart** [comment]
@bardliao @RanderWang @lyakh can you take a look?

@bardliao this would need to be squashed with another patch 13e6f3d6ec67 ('ASoC: codecs: rt*-sdw.c: fix off-by-one port count') since it's really the same issue.

**Comment by bardliao** [comment]
> @bardliao this would need to be squashed with another patch [13e6f3d](https://github.com/thesofproject/linux/commit/13e6f3d6ec679d3cfaf256bd049696552ebf83ce) ('ASoC: codecs: rt*-sdw.c: fix off-by-one port count') since it's really the same issue.

acked.

**Comment by RanderWang** [comment]
@plbossart  a CI compiling error
````C
drivers/soundwire/mipi_disco.c: In function ‘sdw_slave_read_prop’:
drivers/soundwire/mipi_disco.c:292:12: error: unused variable ‘i’ [-Werror=unused-variable]
  292 |  int nval, i, dp0 = 0;
      |            ^
cc1: all warnings being treated as errors
scripts/Makefile.build:280: recipe for target 'drivers/soundwire/mipi_disco.o' failed
````

**Comment by plbossart** [comment]
> @plbossart a CI compiling error
> 
> ```c
> drivers/soundwire/mipi_disco.c: In function ‘sdw_slave_read_prop’:
> drivers/soundwire/mipi_disco.c:292:12: error: unused variable ‘i’ [-Werror=unused-variable]
>   292 |  int nval, i, dp0 = 0;
>       |            ^
> cc1: all warnings being treated as errors
> scripts/Makefile.build:280: recipe for target 'drivers/soundwire/mipi_disco.o' failed
> ```

fixed now

**Pull request merged**